### ISSUE TITLE
Support scan'ing into an existing array

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#20205]
Currently, chapel scans always create a new array. Even when storing the result into an existing array, this results in new array creation followed by an assignment into the existing array. e.g.:

```chpl
var outputArray: [inputArray.domain] inputArray.eltType = 0;
outputArray = + scan inputArray;
```

Will result in creating a temporary array from the `+ scan` and then assigning that into `outputArray`. For use-cases (like a few in Arkouda) where you could reuse an existing array this wastes time allocating, faulting-in, and potentially registering a new array.  

Here I'm wondering if we could support scan'ing into an existing array through a language or library extension or with a compiler optimization. For the compiler optimization I'm wondering if we can detect the patten and just have the scan use the output array, but it's not really clear to me how this would work.

It's also not really clear to me what language level support would look like. For a library you could imagine passing in the result array to a function. Maybe something like `proc exclusiveScan(operator, inputIterable, outputIterable)`.

**Comment by bradcray** [comment]
This issue seems somewhat related to @mppf's musings about having the compiler match compound patterns like `x = y op z;` in [#17999](https://github.com/chapel-lang/chapel/issues/17999#issuecomment-976579379) and subsequent comments, not to mention long-term desire to optimize bigint operations (which are also mentioned on that topic).

**Comment by mppf** [comment]
> Currently, chapel scans always create a new array. Even when storing the result into an existing array, this results in new array creation followed by an assignment into the existing array. e.g.:
> 
> ```chpl
> var outputArray: [inputArray.domain] inputArray.eltType = 0;
> outputArray = + scan inputArray;
> ```
> 
> Will result in creating a temporary array from the `+ scan` and then assigning that into `outputArray`. For use-cases (like a few in Arkouda) where you could reuse an existing array this wastes time allocating, faulting-in, and potentially registering a new array.  

Right but if it were this

```chpl
var outputArray;
outputArray = + scan inputArray;
```

wouldn't `outputArray` only be initialized by the scan with split init? (In principle, this should apply even if `outputArray` has a declared type). Is the split init not available to you for some reason? Like, are you trying to reuse an array for multiple scans or something? If so it would be good to update the example.

**Comment by ronawho** [comment]
> Like, are you trying to reuse an array for multiple scans or something? If so it would be good to update the example.

Yeah, trying to reuse an array across multiple rounds of radix sort was the original motivating example. I updated the example. 

> The spec doesn't have any restrictions on a scan's input and output being the same array, or overlapping slices of the same array. If the compiler enables this as an optimization, would it be safe to apply in those cases? If not, could it detect those cases at runtime and not use the optimization? Or call it out as unsupported in the spec (and make it a runtime halt when checking is on)?

Interesting and good point. I hadn't been thinking about aliasing, but I believe that is a problem for any compiler driven optimization (assuming we don't make it illegal)

**Comment by ronawho** [comment]
This was originally motivated by Arkouda's radix sort, but in thinking about it some more the overhead is pretty negligible in the context of sort. We're scanning an array with only `2**16` elements per task, so we're really only paying the array creation cost and then memory fault-in on 1/2 MB per task, all of which should be almost trivially fast given the other overheads in sort.

I don't want to close this issue, since I think it still has merit, but I don't think it's particularly high priority since sort won't really benefit and there's not any other use cases in Arkouda where there'd be an obvious/natural reason to store into an existing array.
